backtrack/subset.go

```diff
@@ -0,0 +1,38 @@
+// subset.go
+// description: subset of a given set
+// details:
+//In mathematics, set A is a subset of a set B if all elements of A are also elements of B; B is then a superset of A. It is possible for A and B to be equal; if they are unequal, then A is a proper subset of B. The relationship of one set being a subset of another is called inclusion (or sometimes containment). A is a subset of B may also be expressed as B includes (or contains) A or A is included (or contained) in B.
+//  - https://en.wikipedia.org/wiki/Subset
+// author: [sixwaaaay](https://github.com/sixwaaaay)
+// see subset_test.go
+
+package backtrack
+
+func Subset(nums []int) (set [][]int) {
+	n := len(nums)
+	// init index array for record the index of each element selected
+	index := make([]bool, n)
+	for i := 0; i < n; i++ {
+		index[i] = false
+	}
+	var backtrack func(int)
+	backtrack = func(depth int) {
```

Comment:
the backtrack function is recursive, so when I use closure, it must be this form of definition

Reply:
Hmm, then I don't think you need closure in this case - most of the things that you are using through closure are easily passable to the function as arguments by reference. Personally I prefer to use closures only when dealing with iterative changes rather than recursive changes.

Reply:
There is one thing I think is worth considering - haven't really thought about this but maybe this behaviour is not completely in the spec:

```Go
var someFunc func(int)

someFunc = func(i int) { someFunc (i-1) }

go func() {
	someFunc = func(i int) {...}
}()

go func() {
	someFunc = func(i int) {...}
}()

someFunc = somethingElse

```
I don't think it is very clear what will happen to the first someFunc call since one of the go routine is changing the value of the function to `someOtherFunc`.

You can argue that you are not doing that in this code and you'd be right. But this is one of those Gotchas that are very difficult to debug in a large application.

Up to you to change this but this pattern of using closure recursion has a big issues that should not be used in production code - since this is a learning repo, this issue cannot be overlooked because people see this magic happening and decide to use this in interesting ways that can potentially result in weird results.